crates/ibc/src/core/ics24_host/identifier.rs

```diff
@@ -123,46 +100,95 @@ impl ChainId {
 
     /// with_version() checks if a chain_id is in the format required for parsing epochs, and if so
     /// replaces it's version with the specified version
     /// ```
     /// use ibc::core::ics24_host::identifier::ChainId;
-    /// assert_eq!(ChainId::new("chainA".to_string(), 1).with_version(2), ChainId::new("chainA".to_string(), 2));
+    /// assert_eq!(ChainId::new("chainA", 1).with_version(2), ChainId::new("chainA", 2));
     /// assert_eq!("chain1".parse::<ChainId>().unwrap().with_version(2), "chain1".parse::<ChainId>().unwrap());
     /// ```
     pub fn with_version(mut self, version: u64) -> Self {
-        if Self::is_epoch_format(&self.id) {
-            self.id = {
-                let mut split: Vec<&str> = self.id.split('-').collect();
-                let version = version.to_string();
-                if let Some(last_elem) = split.last_mut() {
-                    *last_elem = &version;
-                }
-                split.join("-")
-            };
-            self.version = version;
+        if self.version != 0 {
+            if let (name, Some(_)) = Self::split_chain_id(&self.id) {
+                self.id = format!("{name}-{version}");
+                self.version = version;
+            }
         }
         self
     }
+
+    /// Splits chain_id into name and version if the id includes epoch number.
+    ///
+    /// Chain id with epoch number has format `{name}-{version}` where version
+    /// is a positive integer starting with non-zero digit.  If `chain_id` is in
+    /// that format, returns `(name, Some(version))`; otherwise returns
+    /// `(chain_id, None)`.
+    ///
+    /// **Panics** if chain id is in epoch format but the version overflows
+    /// `u64`.
+    fn split_chain_id(chain_id: &str) -> (&str, Option<u64>) {
+        fn split(chain_id: &str) -> Option<(&str, u64)> {
+            let (name, version) = chain_id.rsplit_once('-')?;
+            if name.is_empty() {
+                return None;
+            }
+            let (car, cdr) = version.as_bytes().split_first()?;
+            if b'1' <= *car && *car <= b'9' && cdr.iter().all(u8::is_ascii_digit) {
+                Some((name, version.parse().unwrap()))
+            } else {
+                None
+            }
+        }
+        match split(chain_id) {
+            Some((name, version)) => (name, Some(version)),
+            None => (chain_id, None),
+        }
+    }
+}
+
+impl Default for ChainId {
+    fn default() -> Self {
+        Self {
+            id: String::from("defaultChainId-1"),
+            version: 1,
+        }
+    }
+}
+
+impl<'a> From<&'a str> for ChainId {
+    fn from(value: &'a str) -> Self {
+        let version = Self::chain_version(value);
+        Self {
+            id: String::from(value),
+            version,
+        }
+    }
+}
+
+impl From<String> for ChainId {
+    fn from(value: String) -> Self {
+        let version = Self::chain_version(&value);
+        Self { id: value, version }
+    }
```

Comment:
Panicking is not acceptable; then one could easily crash a chain by sending a `ClientState` object with a very long version number ([here](https://github.com/cosmos/ibc-rs/blob/cef9b9285cff60e029aeef5e46f26ffdc31942ab/crates/ibc/src/clients/ics07_tendermint/client_state.rs#L640)).

However, our previous solution of storing `0` in that case is also not that great. I suggest we simply implement `TryFrom<String>` and `TryFrom<&'a str>` instead, and bubble the errors up instead of hiding them. In tests, we can just `unwrap()`. Note that returning an error is consistent with [ibc-go's implementation](https://github.com/cosmos/ibc-go/blob/96808ebb72ec8db1edf71f4bff7e8ee68440237d/modules/core/02-client/types/height.go#L177-L181) (where a panic is equivalent to failing to transaction).

Good catch though, I never noticed this behavior before.

Reply:
So to clarify, the desired behaviour is:
- `foo` → valid not in epoch format,
- `foo-bar` → valid not in epoch format,
- `foo-0` → valid not in epoch format (?),
- `foo-1` → valid in epoch format and
- `foo-18446744073709551616` → invalid.

PS. Regarding panicking, I was under the impression that this code is
executed within the VM where panics result in transaction failure.

Reply:
What’s the output of `ChainId::chain_version("foo-18446744073709551616")`?

What’s the output of `ChainId::from(tendermint::chain::Id("foo-18446744073709551616"))`? That conversion is used for serde deserialisation.